### ISSUE TITLE
Simplify hs.geometry by separating its concerns

### ISSUE THREAD

**Issue by cmsj** [Hammerspoon/hammerspoon#776]
"You can create these objects in many different ways" - I think this is a poor API design choice, we should have simple ways to create these simple objects, and not be trying to parse a myriad of options.

I intend on removing `hs.geometry.new()` entirely, and return the point/rect/size/unitrect/etc constructors to creating objects with number arguments, unless I hear excellent arguments why that's a bad idea.

(I realize this is @lowne 's baby, but he is busy with a real life baby right now, so I may delay until he has time to chime in)

**Comment by asmagill** [comment]
Where would this leave us with regards to the idea of allowing an array/table of 2-4 numbers being treated as a point/size/rectangle as discussed re possible changes to the LuaSkin push/get methods mentioned in the 1.0 thread?  Should I still plan on adding this support?

**Comment by lowne** [comment]
Hello there! :)

I'm unsurprisingly against this; my arguments in order of increasing importance:

1. API design philosophy: *blah blah* the API targets Lua and not C, *blah blah* dynamic language interpreters must deal with a "myriad of options" all the time and that's kind of the whole point of being dynamic, *blah blah* more Lua vs C ranting *blah* abstraction layers *blah* [for newcomers that read this and think I'm crazy: 1. I'm indeed not *entirely* sane atm, with the new&improved sleep regimen 2. find another random thread with me in the discussion, good chance you'll find *really* long treatises on the subject]
2. it has reached (after admittedly more bugfixes than I would've liked) that delicate and beautiful state of "it works, as long as nobody touches it" that has blessed unmaintainable codebases since time immemorial, so what's the problem? (but seriously, just to preempt any (unstated so far) opposition on grounds of complexity of the *code* itself (as opposed to the API): it's certainly not the most straightforward and self-explanatory Lua code in existence, but it's not code golf either; there may be some tricky bits here and there (otherwise it would've been born bug-free) but the parser in `.new` isn't amongst them).
3. for more orderly/less anarchic minds, `.point/size/rect/unitrect()` constructors with number arguments *are still there* (unitrect isn't documented, then again rects and unitrects are really the same thing); taking the whole, overenthusiastic intro blob in the module-level docstring and shoving it into `new()`'s, well out of the limelight, is certainly doable and a matter of philosophy as per point 1.
4. (this is the big one) there are **a whole lot of downstream dependencies** all over the place; I fear that going after them will be quite the arduous task; and some things (e.g. the entirety of `hs.window.layout`) will need to be fully eradicated, as there's really no way to rework them into using the 'dumb' (heh) constructors.

[...] I now see that this issue has been prompted by a convo in IRC, which was caused by an error in parsing the docs (😜). I had started to elaborate on point 4, to try and assist with the dire ramifications of axing `new()`, but I think a better solution is simply to obliterate `new()`'s *docstring* and any mention of it in the module docstring.

**Comment by cmsj** [comment]
@lowne 

1. I disagree fundamentally that the dynamic-ness of the language defines how loose it needs to be with its API. We can perfectly well define strict interfaces that are simple for us to deal with, and don't overload users with many ways of doing things. Anything else is just adding complexity for its own sake, and makes testing/debugging/extension a nightmare.
2. I am on a reliability/testing kick at the moment, because our userbase is growing while our reliability is falling. I would rather sacrifice the number of ways we can express a table of 2-4 numbers, to get more maintainable code that is more reliable :)
3. I am definitely going to start by hiding `.new()` since that only involves changing docstrings. From there, the next step is to write tests for the constructors, so we can refactor/rewrite the code more easily. I don't think that's going to be enough though, because...
4. If `hs.window.layout` can't exist without `.new()`, then I am worried about it too. It only uses the function 10 times, and if it's not clear in each of those cases whether we're dealing with a point, size or rect, then that suggests to me that `layout` is also too confusing.

I really can't stress enough, how critical a phase we are in right now. We're getting more and more attention, we've been landing features left and right, but our codebase is almost entirely without tests, and our various parts are becoming very inter-dependent. We *must* now focus heavily on reliability and get ourselves to a state for a 1.0 release. The list of stuff to do is already huge, because we have so much code, and some of it is so old, but if reworking newer stuff also needs to happen, then so be it.

I also want to be clear that I'm not trying to criticise you in any way, you write good quality, idiomatic Lua. The issue here is that I think idiomatic Lua is insane, and I'm sufficiently worried about our reliability that I'm going to start kicking back hard against the dense, twisty maze that is some of our Lua (but not in isolation, there are plenty of bits of crappy non-idiomatic Lua, and even more bits of crappy Objective C, that need to be taken care of).

**Comment by lowne** [comment]
[I advise to skip this bit below and jump directly to **SOLUTIONS** below, the continuation of my previous post]

So the problem seems to be both the API *and* the code.

First some misc observations; this is just me reasoning aloud trying to understand what's the *real* issue here (maybe I'm just overtired, but so far I don't have a clear answer to that), in order to help with the resolution:

1. I'm frankly a bit baffled that a single case of someone *misreading the docs* immediately devolves into "let's eradicate all the code!" instead of what I think would be the logical consequent, "let's make the docs more clear." 
2. I'm even *more* surprised to learn that, since the focus is on improving reliability and stability, you would go about removing code about which we have zero indications of it being buggy or otherwise problematic, but with a ton of downstream dependencies, which can only *decrease* stability (until all the replacement code is in place, has met the real world, and has been in turn thoroughly debugged)
3. The immediate way to increase stability is to extend test coverage, which from what I can gather is being worked on in full swing; it'll be the tests themselves to inform about unreliable code, not the other way around ("this code *looks* unreliable, so it can't be tested properly")
4. Speaking of testing: 
  - *testing* 1 method that does 4 things takes the exact same amount of effort than testing 4 methods that do 1 thing each
  - *extending* the 4-in-1 method is *less* work and results in far less code
  - *debugging* the 4-in1 method is indeed more work because there is more code to begin with: the additional code that decides which of the 4 things to do
5. Getting rid of this extra code might therefore seem an easy win for debugging (but debugging what, in this instance? see point 1), but all it's doing is just shifting the burden downstream to the API users; *some* of the downstream code that used to call into the One Method will simply switch to call the specialised version it needs, but some other code will now have to decide which specialised version it needs. And so we're back to re-adding the same dispatch code, only now it's in 20 different places with 20 slightly different versions...

Point 5 above is just yet another wording of the abstraction principle (I feel justified to rehash it here because it's not at all theoretical - consolidating not-exactly-duplicated code scattered around various modules is *exactly* how the current incarnation of `hs.geometry` was born), and for now I will assume there's no disagreement there; therefore the *real* issue (by exclusion) must simply be that you dislike and/or deem unnecessary only the part of the API that's intended for *human* users, i.e. passing *strings* formatted in various fanciful ways to `hs.geometry.new`.

If that's the case, the solution is fairly straightforward:

## Solutions
#### The problem is the end-user API
- remove the patterns and everything else string-related from the local function `parse` (and disallow strings, ofc)
- rewrite the docs for `hs.window.layout`: it doesn't strictly *need* strings (nor those *much* more horrifying parts of its code that deal with string parsing), it can work just as well with 5-levels-deep tables (tons of fun to type those by hand, so that's a win)

#### The problem is "overloading" in general
[rant about the oddity of *electing* to use a dynamic language to design "strict" interfaces (brrr, the word itself reeks of Java) snipped for brevity and *captatio benevolentiae*]
If we *don't* agree on point 5, things get more hairy, but here's (some of) what needs to be done, from the top of my head:
- remove `geometry.new` altogether
- from `hs.screen` remove `find()` and probably a few more methods; adjust other affected methods to use the specialised constructors
- from `hs.window` remove `find()`, `move()`, etc.; it may or may not be possible to just swap in specialised constructors in other methods dealing with frames/points
- (optionally) add `.findScreenWithResolution()`, `.findScreenContainingPoint()`, `.findScreenContainingRect()`, `.findScreenWithFrame()` methods in `hs.screen`; if omitted, such logic must be implemented for each further-downstream users; alternatively the downstream users can be altogether removed as well
- same as above for `find` and `move` in `hs.window`
- `hs.expose`, `hs.window.switcher`, `hs.window.highlight` need adjusting
- if the "specialisation" is moved downstream all the way (so with all the `hs.screen.find*` etc. replacement methods in place):
  - `hs.window.filter` needs adjusting the parts dealing with screen- and screen region- filters
  - `hs.window.layout` needs a complete rewrite
- conversely, if there's a line in the sand:
  - `hs.window.filter` needs removal of the parts dealing with screen- and screen region- filters
  - `hs.window.layout` has to go

#### The problem is Lua
Finally, if `hs.geometry` warrants further fixing/simplification of twisty code beyond the removal of `.new`:
  - `hs.expose`, `hs.window.switcher`, `hs.window.highlight` also need to go

I'm sure I'm forgetting something but this should be a start.

**Comment by heptal** [comment]
FWIW, I don't think all the string constructors are terrible. But they really need to be cleaned up.
>`"X Y"` or `"X,Y"` creates a point
`"WxH"` or `"W*H"` creates a size
`"X Y/WxH"` or `"X,Y W*H"` (or variations thereof) creates a rect given its width and height from a corner
`"X1,Y1>X2,Y2"` or `"X1 Y1 X2 Y2"` (or variations thereof) creates a rect given two opposite corners
`"[X,Y WxH]"` or `"[X1,Y1 X2,Y2]"` or variations (note the square brackets) creates a unit rect where x=X/100, y=Y/100, w=W/100, h=H/100

In the non-string constructors `hs.geometry(X,Y,W,H)`, `hs.geometry({X,Y,W,H})`, the order of arguments is x, y, width, and height, which is expected since that pattern is in all kinds of languages and frameworks for constructing screen rects. It's a well-known convention. For the arguments to represent x1, y1, x2, and y2, parameters are generally named as in `hs.geometry({x1=X1,y1=Y1,x2=X2,y2=Y2})` because it is a specialized case.

The issue I had was that I tried `"[a,b,c,d]"` as a string constructor for a unitrect, expecting those values to represent x, y, width, and height. This particular pattern was not even documented at all, but to me, intuitively, it should have represented the 'normal' case. (I suppose technically, I didn't misread the docs because this case wasn't actually documented and I went off well-known convention and the existing pattern for `hs.geometry`'s non-string constructors). There is also an error in the docs as `"[X1,Y1 X2,Y2]"` is said to 'create a unit rect where x=X/100, y=Y/100, w=W/100, h=H/100'.

It doesn't make sense, to me, that the x1, y1, x2, y2 constructor is represented by `"X1,Y1>X2,Y2"` AND `"X1 Y1 X2 Y2"` AND `"X1,Y1,X2,Y2"` (and corresponding bracketed forms) since it is rather a special case. I think `"X1,Y1>X2,Y2"` is meaningless and should be removed altogether. `"a,b,c,d"` (and even `"a b c d"`) should probably represent `"X,Y,W,H"` as is the much more common convention. Just search `x y w h rect` vs `x1 y1 x2 y2 rect` for proof of `x y w h` supremacy. 

Perhaps something like `"(X1,Y1),(X2,Y2)"` could represent a pair of points, but it seems to me a rect made from a pair of points should altogether be left to a more explicit constructor like `{x1=X1,y1=Y1,x2=X2,y2=Y2}`. It doesn't seen burdensome to users to have to write a tiny helper function to generate an explicit representation if they happen to love `x1 y1 x2 y2` style rect-making.

Despite this, I think
 >`"X Y"` or `"X,Y"` creates a point
`"WxH"` or `"W*H"` creates a size

is fine because there is more context as you are explicitly creating either a point or a size.

#### In Summary
`"X Y"` or `"X,Y"` -> point seems fine
`"WxH"` or `"W*H"` -> size seems fine
`"a b c d"` or `"a,b,c,d"` -> `x y w h`-style rect seems fine

Use `{x1=X1,y1=Y1,x2=X2,y2=Y2}` constructor for `x1 y1 x2 y2`-style rects
Remove  `"X1,Y1>X2,Y2"` and `"X Y/WxH"` and  `"X,Y W*H"` altogether

I don't believe this simplification would require significant downstream changes (if any), and it seems much more expressive than the current complexity we have (to me at least).

**Comment by lowne** [comment]
The docs would definitely benefit from clarification (and perhaps a more formal structure, instead of a bunch of examples and then let the reader pattern-match to fill in the gaps), starting with the fact that rects and unitrects are the same thing; and that the '[]' string syntax, far from being mandatory for unitrect creation, is a mere convenience.
`'X,Y WxH'` - or just `{x,y,w,h}` - where x,y,w,h are in [0,1] will create a "unitrect" (which, again, is really just a rect) just fine:
```
> geometry.type(geometry'0,0 1x1')
unitrect
```
As for x,y,w,h vs x1,y1,x2,y2, I agree that it's unusual, and in fact as you noted a naked table is interpreted as {x,y,w,h}, whereas x1 etc needs named keys (being the "exception"); but here (talking about strings) the rationale is consistency:
- a point is `'X,Y'`, and a size is `'WxH'`
- if a rect is defined by a point+a size, then it should be constructed via a point 'X,Y' + a size 'WxH'
- if a rect is defined by two points (two corners), then it should be constructed via point 1 'X1,Y1' + point 2 'X2,Y2'

In fact even if the string notation for (or the whole notion of) x1,y1,x2,y2 rects were removed, I still think the xywh string notation should remain as is, unless notations for points and sizes also change accordingly.

Maybe, to strongly differentiate the "string" use case (more on that below) from the *regular* way of creating geometry objects (**tables**) - and also deal with the other confusing syntaxes mentioned - the syntax for rects (or unitrects) could be limited to
- "X Y WxH"
- "X1 Y1 > X2 Y2"

so, no commas, and every space etc mandatory.

At any rate, strings weren't meant to be the primary way of constructing geometry objects in user scripts, but for downstream - specifically `hs.window.layout`, which can use one long string (containing several "inline" geometry objects) per rule; therefore that part of the module-level docstring could be de-emphasised or just moved over to `hs.window.layout`.

**Comment by heptal** [comment]
In this context, people will tend to grok a 4-element 'tuple' of the form `A,B,C,D` as a rect described by `x,y,width,height` because that convention is incredibly widespread. They will also, separately and easily, grok `A,B` as your standard `X,Y` point and interpret `AxB`/`A*B` as a `WxH` size. 

The 'consistency' you are arguing for is flawed because it is not consistent with people's existing mental models. It is just the mechanical gluing of these syntactic forms together without considering semantics. If asked to represent a pair of distinguishable points, on paper or in code, most people would choose a form like `(x1,y1), (x2,y2)` over `(x1,y1,x2,y2)`. Whether they use parens or braces or something else is irrelevant. Nobody would use something like `"X1 Y1 > X2 Y2"` or `X1,Y1 > X2,Y2`. The > symbol is meaningless and arbitrary here. 

Similarly, they would expect to be able to represent a full rect as `X,Y,W,H` because it's very well established (see CGRect, canvas rect, Qt rect, Windows rect, and basically almost any other decent object-oriented UI library). It is nonintuitive to represent it as `X Y/WxH`, `X,Y W*H`or whatever; nobody expects that form. This 'mechanical consistency' reduces expressiveness. The reader shouldn't be forced to 'pattern-match' over foreign syntax examples at all.

The reason for the string constructor existence is brevity and simplicity. Imposing additional restrictions like mandatory spaces and no commas works against that purpose. The absolute clearest choice for a rect is `X,Y,W,H`.

Thinking about it some more I think the `X Y` point constructor should be removed as well, leaving only `X,Y`. Spaces are an odd choice; we are not using LISP. Commas impose no additional burden or reduce expressivity. For size, both `WxH` and `W*H` are acceptable because they are established.

Here are what I think should be the available string constructors, which should exist for _convenience_ and not as a platform for code art:
- `"X,Y"` -> `point` 
- `"WxH"` or `"W*H"` -> `size`
- `"X,Y,W,H"` ->  `rect` (and `unitrect` with [0,1] ranges)

All of our `hs.geometry` objects can be expressed just fine with this 'limited' shorthand.
No bracket syntax.
No mixed point/size strings.
No points glued together with >.
No spaces.
Docs are simplified.
Code is more robust and simplified.

It's really not a burden to make explicit tables (or a helper function) if you wish to describe rects using a pair of points.

**Comment by cmsj** [comment]
I'm currently thinking that a suitable strategy for reworking this, would be:

1. Change the `hs.geometry` docs to tell people to use `.rect()`/`.size()`/`.point()`/etc
2. Remove `.new()`
3. Refactor the string parsing into a (documented, but discouraged) `hs.geometry.newFromString()`
4. Reduce the number of supported string formats to something approximating the list in @heptal 's comment
5. Write tests for all of hs.geometry

Users get a simple, API that minimises the risk of getting confused by the number of options, downstream users can support geometries inside strings without having to have their own parsing gunk.

Is there anything wrong with this plan?

**Comment by cmsj** [comment]
I'm going to close this because I haven't touched it in a year and I don't seem to care as much as I used to :)
